soc/nordic/nrf54h/poweroff.c

```diff
@@ -0,0 +1,182 @@
+	}
+
+	/* Disable retention */
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+
+	k_cpu_idle();
+}
+
+void z_sys_poweroff(void)
+{
+	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO, 0);
+	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, false);
+
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+
+	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_SYSTEMOFFREADY);
+
+	suspend_common();
+
+	CODE_UNREACHABLE;
+}
+
+__weak void z_pm_soc_resume(void)
+{
+}
+
+/* Resume domain after local suspend to RAM. */
+void z_pm_sys_resume(void)
+{
+	if (IS_ENABLED(CONFIG_ICACHE)) {
+		/* Power up and re-enable ICACHE */
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_ICACHE, true);
+		sys_cache_instr_enable();
+	}
+
+	if (IS_ENABLED(CONFIG_DCACHE)) {
+		/* Power up and re-enable DCACHE */
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_DCACHE, true);
+		sys_cache_data_enable();
+	}
+
+	/* Re-enable domain retention. */
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+
+
+	z_pm_soc_resume();
+}
+
+/* Function called during local domain suspend to RAM. */
+int z_pm_sys_suspend(void)
+{
+	/* Set intormation which is used on domain wakeup to determine if resume from RAM shall
+	 * be performed.
+	 */
+	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO,
+					  NRF_RESETINFO_RESETREAS_LOCAL_UNRETAINED_MASK);
+	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+
+	suspend_common();
+
+	/*
+	 * We might reach this point is k_cpu_idle returns (there is a pre sleep hook that
+	 * can abort sleeping.
+	 */
+	return -EBUSY;
+}
+
+void pm_s2ram_mark_set(void)
+{
+	/* empty */
+}
+
+bool pm_s2ram_mark_check_and_clear(void)
+{
+	bool unretained_wake;
+	bool restore_valid;
+
+	unretained_wake = nrf_resetinfo_resetreas_local_get(NRF_RESETINFO) &
+			  NRF_RESETINFO_RESETREAS_LOCAL_UNRETAINED_MASK;
+	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO, 0);
+
+	restore_valid = nrf_resetinfo_restore_valid_check(NRF_RESETINFO);
+	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, false);
+
+	return (unretained_wake & restore_valid) ? true : false;
+}
+
+#if CONFIG_PM
+
+__weak bool z_pm_do_suspend(void)
+{
+	return true;
+}
+
+void do_suspend(void)
+{
+	if (!z_pm_do_suspend()) {
+		return;
+	}
+
+	/*
+	 * Save the CPU context (including the return address),set the SRAM
+	 * marker and power off the system.
+	 */
+	(void)pm_s2ram_suspend(z_pm_sys_suspend);
```

Comment:
It seems this function should be called with IRQs disabled (wrapped with `__disable_irq()` - `__enable_irq()`).

If IRQs are enabled, we risk a race condition: after the NVIC or MPU state is saved in RAM, ZLI is triggered, resulting in a change in NVIC or MPU. When ZLI ends, this function resumes execution storing CPU registers, configuring LRCCONF, etc. When the CPU wakes up and restores NVIC and MPU, it restores incorrect values, saved before ZLI was triggered.

If IRQs are disabled, ZLI wouldn't be handled until we exit this function, `WFI` would result in `NOP`, so the system state would be coherent.